### ISSUE TITLE
Stateless Transformation

### ISSUE THREAD

**Issue by n2cholas** [deepmind/optax#104]
Right now, there's some boilerplate for defining simple gradient transformations. If a user wants to implement custom weight decay, clipping, constraints, etc without any state, they still have to define a nested function with an `init` that does nothing, and handle the empty state.

I think it'd be convenient to provide a `stateless` transformation that accepts a function to apply to the updates and params. We can also do the `jax.tree_multimap` for the user by default.

```python
weight_decay = optax.stateless(lambda g, p: g + 0.1 * p)
```

If the user wants to define a function that does the tree_multimap themselves:

```python
def my_function(updates, params):
    return jax.tree_multimap(lambda g, p: ..., updates, params)

optim = optax.stateless(my_function, on_leaves=False)
```

In my view, this is a very clean way to implement simple stateless transformations. I think the `on_leaves` argument could use a better name, though. I'd be happy to implement this if it sounds reasonable.

**Comment by mtthss** [comment]
This is an interesting suggestion. Lets discuss our options here a bit more.

Some initial thoughts:

One concern for me is the fact that `optax.stateless` suggests that it handles the empty state,
but it does not immediately suggest that it would do the tree mapping for me as well.

While `on_leaves` allows me to disable it,
it's weird to have to disable something that its not clear was happening in the first place.

Not the end of the world ofc, maybe we can just find a different name than `stateless`?
An alternative would be to disentangle the two aspects: statelessness and tree_mapping,
but maybe `optax.stateless(optax.map(fn)` is a bit two cumbersome.

In any case we would probably want to make this (or these) decorators so that you can do things like
```
@optax.stateless
@optax.tree_map  # if we don't incorporate it in stateless directly
def user_fn(g, p):
  return g - 0.1*p
```

**Comment by n2cholas** [comment]
Good point @mtthss, I agree it's more transparent to make the `tree_multimap` opt-in instead of the default. I think making `on_leaves=False` by default is a good way to achieve this.

Letting the user do the `tree_multimap` themselves could work too (not `tree_map`, since we're tree-mapping over both updates and parameters). I don't think the following is _too_ bad

```python
tx = optax.stateless(optax.tree_multimap(lambda g, p: g - 0.1*p))
```

Though perhaps a bit clunky, like you said. It's not much better than the user just doing that in native JAX:

```python
from functools import partial
tx = optax.stateless(partial(jax.tree_multimap, lambda g, p: g - 0.1*p))
```

Also, I don't think `optax.stateless` can be a decorator, since it returns a `GradientTransformation` and not a callable.

<hr>

A bit tangential, but I think it's considering alternative use cases for `optax.tree_multimap` before creating the API. In particular, it may be useful for optax to provide a specialized multimap for transformation `update_fn`s that lets you write it for each parameter instead of for the pytree. For example, the `update_fn` for [`optax.trace`](https://github.com/deepmind/optax/blob/master/optax/_src/transform.py#L65-L72) becomes much cleaner:

```python
  @optax.multimap_update  # different name so users don't confuse it with a normal multimap
  def update_fn(update, state, param=None):
    del params
    new_trace = update + decay * state.trace
    new_update = update + decay * new_trace if nesterov else new_trace
    new_trace = new_trace.astype(accumulator_dtype)
    return updates, TraceState(trace=new_trace)
```
This `multimap_update` would have to do a few things:

- Not map over `params` if it is None
- `tree_transpose` the state so it is a pytree of `OptStates` (instead of an `OptState` of pytrees).
- `tree_transpose` the new returned state back into an `OptState` of pytrees.

I can open a new issue for this if you want to further discuss it.

**Comment by mkunesch** [comment]
I really like the `optax.stateless` suggestion! As you said, it would make defining simple transformations much easier.

An additional consideration would be how hyperparameters should be handled, e.g. for the `weight_decay` parameter in `add_decayed_weights`. We would need a function that creates a gradient transformation so that we can pass it e.g. to `inject_hyperparams`.

With your current suggestion I think this would be:

```
def add_decayed_weights(weight_decay):
  return optax.stateless(lambda g, p: g + weight_decay * p, on_leaves=True)
  
opt = add_decayed_weights(weight_decay=0.1)
```

If we make optax.stateless a decorater that directly returns a function that creates a gradient transformation (passing all the hyperparameters through) it would be:

```
add_decayed_weights = optax.stateless(lambda g, p, weight_decay: g + weight_decay * p, on_leaves=True)
opt = add_decayed_weights(weight_decay=0.1)
```

(We'd probably want a different name for `stateless` in this case.)

The latter is a bit shorter and would work as a simple decorator, but I think the former (your suggestion) is easier to understand and more readable.

So yes, I think `optax.stateless` as you proposed with `on_leaves=False` by default sounds great to me!

**Comment by n2cholas** [comment]
Awesome, thanks @mkunesch! I'll get started on a PR.

**Comment by mtthss** [comment]
Great, thanks @n2cholas! And apologies for the slow turnaround on this issue.

**Comment by mkunesch** [comment]
Thanks a lot @n2cholas! Also, I didn't get back to your separate point about `@optax.multimap_update`. 

I agree that this would simplify quite a few optimizers, but I think there are two potential downsides:

1. it could cause confusion when update functions across optax or user code sometimes operate on trees and sometimes on leaves.
2. for an arbitrary state I think the mapping could become confusing (as it differs between updates and state).

In my opinion, for stateless transforms the benefits of being able to define an optimizer in just one line outweigh the potential downside of the confusion (especially since there is no state so point 1 above doesn't apply). However, in the general case for optimizers that have non-trivial state where we have already added all the boiler plate of init and update functions I am not sure whether the benefit of being able to define the update functions on leaves outweigh the downside.

In short, I agree that it would be great to have the `on_leaves` option for stateless transforms, but I'm not sure we should also have `@optax.multimap_update` in general. What do you think?

**Comment by mtthss** [comment]
+1 to @mkunesch analysis

**Comment by n2cholas** [comment]
No more action needed on your side @mtthss, I'll work on this over the weekend!
